torch/csrc/distributed/rpc/rref_context.h

```diff
@@ -109,6 +127,26 @@ class RRefContext {
       std::unordered_set<ForkId, ForkId::Hash>,
       RRefId::Hash>
       forks_;
+
+  // The follow two maps keep UserRRefs alive by holding a shared_ptr to the
+  // RRef instances. A UserRRef must be added into this map if any of the
+  // following two conditions is ture:
+  //
+  // (1) A UserRRef has not been accepted by owner yet.
+  //
+  //     It can be used or shared, but cannot be deleted, and hence kept alive
+  //     in this map. A message of type RREF_USER_ACCEPT will remove the
+  //     corresponding RRef from this map.
+  std::unordered_map<ForkId, std::shared_ptr<RRef>, ForkId::Hash> pendingUsers_;
+
+  // (2) A UserRRef has forked a child UserRRef which has not been accepted by
+  //     the owner yet.
+  //
+  //     In this case, this UserRRef cannot send out RREF_USER_DELETE message,
+  //     as it could potentially trigger the OwnerRRef been deleted before the
+  //     owner learns about the forked child.
+  std::unordered_map<ForkId, std::shared_ptr<RRef>, ForkId::Hash>
+      pendingChildren_;
```

Comment:
Curious about the reason for us to keep two maps: pending user and pending children. From caller point of view, it just needs to hold on to a UserRRef until it hears back from the callee. If the callee is owner, it's pendingUsers. If the callee is not owner, it's pendingChildren_. The ack mechanism is a bit different, but that doesn't mean we need to differentiate the mechanism to keep rref alive?

Reply:
Because they hold RRefs for different reasons. 

1. `pendingChildren_` holds `RRef` (user or owner) when the caller hasn't received the ack from the  rpc/remote **callee**.
2. `pendingUser_` holds and `RRef` (user only) when it has not received the confirmation from the **owner**.

Store them in separate maps might help us craft more accurate error messages when we have RRef leaks (not in the code yet). But I agree with you that technically, they can be merged into one, as every user/fork will have their unique fork id. Do you prefer to keep them separate or consolidate into one? @gqchen

Reply:
My choice would be calling both of them "pendingRRefs_". "pendingRRefs_" will be updated accordingly depending on different scenarios. This will reduce the future code maintenance overhead. Nevertheless, this is not a blocker.

Reply:
I agree with @gqchen but it is very important to provide very specific error messages. Would it make sense to store tracing information along with this map to ensure that the error message contains all the information needed to debug an infrastructure instability? [doesn't need to be in this PR]

Reply:
Sure. Tracing message can contain who is the parent of this RRef. What else should be put in tracing info?